src/lib/utils/colors.ts

```diff
@@ -1,4 +1,129 @@
-import { isNumber } from "./types";
+import {
+  TonalPalette,
+  argbFromHex,
+  hexFromArgb,
+  themeFromSourceColor,
+  type Theme,
+} from "@material/material-color-utilities";
+import { isNumber } from "./types.js";
+import { convertCase } from "./string.js";
+
+export type HexValue = `#${string}`;
+
+export type QuaffColors = {
+  primary: HexValue;
+  "on-primary": HexValue;
+  "primary-container": HexValue;
+  "on-primary-container": HexValue;
+  secondary: HexValue;
+  "on-secondary": HexValue;
+  "secondary-container": HexValue;
+  "on-secondary-container": HexValue;
+  tertiary: HexValue;
+  "on-tertiary": HexValue;
+  "tertiary-container": HexValue;
+  "on-tertiary-container": HexValue;
+  neutral: HexValue;
+  "on-neutral": HexValue;
+  "neutral-container": HexValue;
+  "on-neutral-container": HexValue;
+  "neutral-variant": HexValue;
+  "on-neutral-variant": HexValue;
+  "neutral-variant-container": HexValue;
+  "on-neutral-variant-container": HexValue;
+  error: HexValue;
+  "on-error": HexValue;
+  "error-container": HexValue;
+  "on-error-container": HexValue;
+  surface: HexValue;
+  "surface-dim": HexValue;
+  "surface-bright": HexValue;
+  "on-surface": HexValue;
+  "on-surface-variant": HexValue;
+  "surface-container-lowest": HexValue;
+  "surface-container-low": HexValue;
+  "surface-container": HexValue;
+  "surface-container-high": HexValue;
+  "surface-container-highest": HexValue;
+  "inverse-surface": HexValue;
+  "inverse-on-surface": HexValue;
+  "inverse-primary": HexValue;
+  outline: HexValue;
+  "outline-variant": HexValue;
+  scrim: HexValue;
+  shadow: HexValue;
+};
+
+export function generateColors(from: string): { light: QuaffColors; dark: QuaffColors } {
+  const argb = argbFromHex(from);
+  const palettes = themeFromSourceColor(argb).palettes;
+
+  return {
+    light: getColors(palettes, "light"),
+    dark: getColors(palettes, "dark"),
+  };
+}
+
+function getColors(palettes: Theme["palettes"], mode: "light" | "dark") {
+  const tones =
+    mode === "light"
+      ? {
+          base: 40,
+          onBase: 100,
+          baseContainer: 90,
+          onBaseContainer: 10,
+        }
+      : {
+          base: 80,
+          onBase: 20,
+          baseContainer: 30,
+          onBaseContainer: 90,
+        };
+
+  const getColor = (color: string, palette: TonalPalette) => {
+    color = convertCase(color, "pascal", "kebab");
+
+    return [
+      [color, palette.tone(tones.base)],
+      [`on-${color}`, palette.tone(tones.onBase)],
+      [`${color}-container`, palette.tone(tones.baseContainer)],
+      [`on-${color}-container`, palette.tone(tones.onBaseContainer)],
+    ];
+  };
+
+  const results: Record<keyof QuaffColors, number> = Object.fromEntries(
+    Object.entries(palettes)
+      .map(([color, palette]) => getColor(color, palette))
+      .flat(1)
+  );
+
+  results["surface"] = palettes.neutral.tone(mode === "light" ? 98 : 6);
```

Comment:
Where do these numbers like `98` and `6` come from?

Reply:
MD3 specs. They give the base color and the tone for all colors. I'll give you the link later.

Reply:
Should we have them as constants with meaningful name in an external file?